compiler/rustc_ast/src/ast.rs

```diff
@@ -1423,10 +1423,10 @@ pub enum ExprKind {
     While(P<Expr>, P<Block>, Option<Label>),
     /// A `for` loop, with an optional label.
     ///
-    /// `'label: for pat in expr { block }`
+    /// `'label: for await? pat in iter { block }`
     ///
     /// This is desugared to a combination of `loop` and `match` expressions.
-    ForLoop(P<Pat>, P<Expr>, P<Block>, Option<Label>),
+    ForLoop { pat: P<Pat>, iter: P<Expr>, body: P<Block>, label: Option<Label>, kind: ForLoopKind },
```

Comment:
Instead of making the `await` part of the `for`, I think I would prefer we made it part of the `Pat` in the AST and reject it later in any pattern that isn't in a for loop. The current syntax implies that `let await foo = future;` would be valid. It isn't but we should likely handle it more gracefully. Particularly if there's *any* possibility that we'll allow people to write `for Struct { await foo } in stream` or `match foo { Some(await foo) => {}, None => {} }`.

Reply:
I think I disagree with this approach.

I don't think that we need to tangle up parsing `async PAT` in this PR. First of all, it makes lowering of `for await` in the HIR more difficult, since it makes the "is this a `for await` pattern" condition slightly more non-local (we need to peel off an async pat), and requires complicating the lowering of patterns with a new unreachable arms and/or error recovery.

Second, we discussed at length in a recent wg-async meeting that the idea of a `async PAT` syntax is not desirable at the moment. We also discussed at length whether we even *want* to call it `for await`, instead of something like `async for`, `for async`, etc. If we were to move to that, we'd have to undo all of the `PatKind::Async` here, too.

Last, if we later find that people are accidentally writing `let await PAT = EXPR;` enough that diagnostics should mention them specifically, then we can handle the recovery for that typo separately and without having to represent it in the AST. Simply eagerly erroring in `parse_pat_with_range_pat` if `may_recover && check_kw(kw::Await)` would be simple enough, I think.

Reply:
@eholk's comment above (https://github.com/rust-lang/rust/pull/118847#issuecomment-1857000155) discusses a bit about how `for await` shouldn't be conceptually thought of as `for (await PAT)` but as a totally separate kind of loop, and I tend to agree with that.

I do think that this PR should stay on the simple side -- I find that unnecessarily making parsing more general to carve out space for AST that doesn't exist is a good way to introduce new bugs, and make things like exhaustively matching on `PatKind`s that are only expected on an error path a bit more tedious.

Reply:
(also, I can also see some issues with pattern macro fragments either not matching `async PAT` correctly, or going from pass->fail if we're not careful, because currently `async` is not allowed in `nonterminal_may_begin_with` for pat fragments)